### ISSUE TITLE
[Flow] Click-to-definition doesn't work reliably on JSX Element

### ISSUE THREAD

**Issue by skevy** [facebook/nuclide#353]
First of all - loving the new and and improved OSS Nuclide. Really great work.

When I first installed it, I was dev-ing on a React Native project that uses normal CommonJS modules, and click-to-definition basically worked right out of the box (obviously after I added a flowconfig). You can see an example of that here:

![Working JSX click-to-definition](http://s.sk3vy.com/f2x4/Screen%20Recording%202016-02-04%20at%2010.07%20AM.gif)

However, when I'm dev-ing on a project that uses Haste modules, while I can click-to-definition on the actual `import`, I can't do it on the JSX element:

![Not working JSX click-to-definition](http://s.sk3vy.com/f3cQ/Screen%20Recording%202016-02-04%20at%2010.09%20AM.gif)

Happy to provide you all with any logs or anything you need to help repro the issue.

Fwiw, I can confirm that `flow type-at-pos` is NOT returning a definition on the project where JSX click-to-definition doesn't work. So, in light of that, I presume this is a Flow bug...but I thought I would report it here first to see if any of you were aware of the issue. I know you all use haste almost exclusively internally, so honestly I was expecting the behavior I'm seeing to be opposite of what it actually is :)

**Comment by ssorallen** [comment]
> when I'm dev-ing on a project that uses Haste modules

Did you set Flow to use the Haste module system as well? It's documented in [Flow's `[options]`](http://flowtype.org/docs/advanced-configuration.html#options):

> * `module.system` (`node` | `haste`): the module system to use to resolve `import` and `require`. Haste is used in React-Native.

**Comment by skevy** [comment]
@ssorallen yes. `.flowconfig` (on both projects) is:

```
[ignore]

# We fork some components by platform.
.*/*.web.js
.*/*.android.js

# Some modules have their own node_modules with overlap
.*/node_modules/node-haste/.*

# Ugh
.*/node_modules/babel.*
.*/node_modules/babylon.*
.*/node_modules/invariant.*

# Ignore react and fbjs where there are overlaps, but don't ignore
# anything that react-native relies on
.*/node_modules/fbjs/lib/fetch.js
.*/node_modules/fbjs/lib/ExecutionEnvironment.js
.*/node_modules/fbjs/lib/ErrorUtils.js

# Flow has a built-in definition for the 'react' module which we prefer to use
# over the currently-untyped source
.*/node_modules/react/react.js
.*/node_modules/react/lib/React.js
.*/node_modules/react/lib/ReactDOM.js

# Ignore commoner tests
.*/node_modules/commoner/test/.*

# See https://github.com/facebook/flow/issues/442
.*/react-tools/node_modules/commoner/lib/reader.js

# Ignore jest
.*/node_modules/jest-cli/.*

[include]

[libs]
node_modules/react-native/Libraries/react-native/react-native-interface.js

[options]
module.system=haste

esproposal.decorators=ignore
esproposal.class_static_fields=enable
esproposal.class_instance_fields=enable

module.name_mapper='^image![a-zA-Z0-9$_-]+$' -> 'GlobalImageStub'
module.name_mapper='^[./a-zA-Z0-9$_-]+\.png$' -> 'RelativeImageStub'

munge_underscores=true

suppress_type=$FlowIssue
suppress_type=$FlowFixMe
suppress_type=$FixMe

suppress_comment=\\(.\\|\n\\)*\\$FlowFixMe\\($\\|[^(]\\|(\\(>=0\\.\\(1[0-4]\\|[0-9]\\).[0-9]\\)? *\\(site=[a-z,_]*react_native[a-z,_]*\\)?)\\)
suppress_comment=\\(.\\|\n\\)*\\$FlowIssue\\((\\(>=0\\.\\(1[0-4]\\|[0-9]\\).[0-9]\\)? *\\(site=[a-z,_]*react_native[a-z,_]*\\)?)\\)? #[0-9]+
suppress_comment=\\(.\\|\n\\)*\\$FlowFixedInNextDeploy
```

**Comment by skevy** [comment]
Also, I'm using Flow 0.21.0.

**Comment by ssorallen** [comment]
If `flow type-at-pos` is not returning what you expect, then the problem is likely with Flow. Nuclide makes that same call to Flow and uses the response to jump to the correct location. Can you open an issue with Flow? https://github.com/facebook/flow/issues

**Comment by nmote** [comment]
Thanks for the detailed info. Just want to confirm that you mean `flow get-def`, not `flow type-at-pos`.

**Comment by skevy** [comment]
@nmote I definitely mean type-at-pos...I have not updated to flow .22 yet...but was going to check if it was fixed in that version, given all the updates to react support.

**Comment by nmote** [comment]
`type-at-pos` is unrelated to click to definition. It just returns the type of the thing at the given position. Nuclide's click-to-symbol is backed by `get-def`, so it would make more sense to investigate whether `get-def` returns what you expect.

**Comment by skevy** [comment]
Ok - so now, with Flow 22, and I call get-def, I get:

```
flow get-def Profile/Profile.js 222 15
/private/tmp/flow/flowlib_26f8a380/react.js:148:1,225:7
```

Any thoughts on that? Again, I can post in the Flow issue tracker...but just wasn't sure if anyone on the Nuclide team might have seen this before when testing with a RN project.

**Comment by nmote** [comment]
This slipped through the cracks, sorry. Just found it again while triaging another report. This is indeed an issue with Flow. Please report it on https://github.com/facebook/flow if you would like to follow up.

**Comment by vaukalak** [comment]
@nmote I'm sorry but I can't understand what exactly should we report to flow :)

**Comment by nmote** [comment]
Nevermind, it turns out that it's fixed in Flow master and the fix will be released with Flow v0.31.0.
